### DISCUSSION TITLE
Provide more documentation on how to write custom components (data() vs props, events, etc.)

### DISCUSSION THREAD

**Discussion by nlgranger** [zauberzeug/nicegui#3627] [Ideas / Feature Requests]
### Question

There is no documentation on how to write custom components, just a few examples without comments.
Here are a few questions I collected while trying to implement a component:

- In JS, what is the purpose of data() compared to props ?
- What happens when the web page is refreshed:
  - Which variables and states are restored ?
  - Which methods are called (mounted, ...) ?
- What happens in JS when a prop is modified from the python side.
- In JS: what are events or hooks to monitor changes of props from python.
- When is the template evaluated, which variables are visible to it ?
- What is the recommended way to push events and values from javascript to python ?

**Comment by rodja** [comment]
Thanks for bringing this up @nlgranger. The documentation on how to write custom components is a bit sparse. A full-body tutorial and general explanations would be really good. I'll update the title and change this into a feature request. But I'm not sure if we can find the time to give it the required attention. Maybe someone from the community can help?

**Comment by falkoschindler** [comment]
By the way, you might find answers to most of these questions in a Vue tutorial. Their [documentation](https://vuejs.org/guide/introduction.html) is also pretty good.

**Reply by nlgranger** [reply]
Thanks for the pointer!
